### ISSUE TITLE
Incorrect note for expected type in E0271 when using an associated type

### ISSUE THREAD

**Issue by ohadravid** [rust-lang/rust#64760]
With the following code:
```
#[derive(Debug)]
struct Data {}


fn do_stuff<'a>(iterator: impl Iterator<Item = &'a Data>) {
    for item in iterator {
        println!("{:?}", item)
    }
}


fn main() {
    let v = vec![Data {}];
    
    do_stuff(v.into_iter());
}
```

the diagnostic (in nightly & stable) wrongly complains about the expected type:
```
error[E0271]: type mismatch resolving `<std::vec::IntoIter<Data> as std::iter::Iterator>::Item == &Data`
  --> src/main.rs:15:5
   |
5  | fn do_stuff<'a>(iterator: impl Iterator<Item = &'a Data>) {
   |    --------                             --------------- required by this bound in `do_stuff`
...
15 |     do_stuff(v.into_iter());
   |     ^^^^^^^^ expected struct `Data`, found &Data
   |
   = note: expected type `Data`
              found type `&Data`
```

It should be the reversed.
This makes it difficult to understand that the `into_iter` should be changed to `iter`.

**Comment by estebank** [comment]
CC https://github.com/rust-lang/rust/issues/53965
